Identifier Forest Rewrite

BruJu opened the pull request:
(with many inspiration from https://github.com/BruJu/wasmify-sophia/pull/9 )

- Better defined API (a proper trait has been created: `MaybeTree4`)
- Block is now just a wrapper with a different set order (no more BlockOrder thing)
- Trees chosen at compile time `OnceTreeSet`
- Instantiated trees order is still at execution time thanks to `DynamicOnceTreeSet`.
- `IndexingForest4` is still there to pick OnceTreeSets at runtime
- Forests with trees chosen at compile time `CTForest`
- The forest (the thing the end user is using) is just a `MaybeTree4` implementation which happens to cheat because it has more than 1 tree.
- Forest4 trait
- Ensemblist operations for WasmTree ( https://github.com/BruJu/WasmTreeDataset ) ; maybe the MaybeForest4 and BinaryMaybe4TreeOperations traits should be merged (and naive default implementation should be provided for BinaryMaybe4TreeOperations methods)

***TODO: before merging***
- Cleanup (choose some more meaningful names, ...)
- Documentation
- More tests

***In the future?***
- Triples
- SyncOnceTreeSet, SyncDynamicOnceTreeSet, SyncIndexingForest4

BruJu committed:
New block structure

BruJu committed:
A trait for OnceCell trees?

BruJu committed:
identifier_forest: identifier, position, order

BruJu committed:
tree trait

BruJu committed:
tree predefined

BruJu committed:
new tree enum generator, new forest implem

BruJu committed:
order module

BruJu committed:
tree module

BruJu committed:
Compile time trees

BruJu committed:
Tests in order module, proper module for run time forests, doc for ct forests

BruJu committed:
Forest4 trait, integration of new if API to other crates

pchampin commented on identifier-forest/src/order/_order.rs:
```diff
@@ -0,0 +1,588 @@
+/// Number of different positions to specify in a [`FixedOrder4`].
+///
+/// It correspond to the number of elements in an identifier quad (4).
+pub const ORDER4_NB_OF_TERMS: usize = 4;
```
Is this constant really useful? Since many parts of the code are *structurally* dependant it being 4 (e.g. the 4 parameters A, B, C, D of FixedOrder4) and since the number 4 occurs in many type or constant names in this module, I think that replacing this constant with "4" everywhere it occurs would make the code more readable, without any loss in generality or maintainability.

BruJu replied:
I removed the constant in 62bb9a1b89ccd86b757f5a81e1dbb2ce9929aeca 

It was mostly tentative to document the crate and help implementing a triple forest in the future but I agree that it ended up being code bloat.

pchampin commented on identifier-forest/src/compile_time_forest/mod.rs:
```diff
@@ -0,0 +1,398 @@
+        let best = self.best_conformance(true, &pattern).0;
+
+        match best {
+            2 => ensure_existance(&self, &self.tree2),
+            3 => ensure_existance(&self, &self.tree3),
+            4 => ensure_existance(&self, &self.tree4),
+            5 => ensure_existance(&self, &self.tree5),
+            6 => ensure_existance(&self, &self.tree6),
+            _ => {} /* noop */
+        }
+    }
+
+}
+
+
+/// If best_btree exists, extract the quads that match the pattern
+///
+/// If it doesn't, takes the first btree of this, fills best_btree and then
+/// extract the quads.
+fn ensure_existance<'a, I, MT, Order1, Order2, Order3, Order4, Order5, Order6>(
```
```suggestion
            2 => ensure_existence(&self, &self.tree2),
            3 => ensure_existence(&self, &self.tree3),
            4 => ensure_existence(&self, &self.tree4),
            5 => ensure_existence(&self, &self.tree5),
            6 => ensure_existence(&self, &self.tree6),
            _ => {} /* noop */
        }
    }

}


/// If best_btree exists, extract the quads that match the pattern
///
/// If it doesn't, takes the first btree of this, fills best_btree and then
/// extract the quads.
fn ensure_existence<'a, I, MT, Order1, Order2, Order3, Order4, Order5, Order6>(
```

pchampin commented on identifier-forest/src/compile_time_forest/mod.rs:
```diff
@@ -0,0 +1,398 @@
+            .enumerate()
+            .filter(|v| v.1.is_some())
+            .max_by_key(|v| v.1.unwrap())
+            .map(|t| (t.0 + 1, t.1.unwrap()))
```
```suggestion
            .filter_map(|(i, opt)| opt.map(|score| (i+1, score)))
            .max_by_key(|(_, score)| *score)
```
a little more compact and (IMO) idiomatic

BruJu committed:
Benches from PAC's new_api branch

BruJu committed:
Update identifier-forest/src/compile_time_forest/mod.rs

Co-authored-by: Pierre-Antoine Champin <pchampin@liris.cnrs.fr>

BruJu committed:
Update identifier-forest/src/compile_time_forest/mod.rs

Co-authored-by: Pierre-Antoine Champin <pchampin@liris.cnrs.fr>

BruJu committed:
Remove ORDER4_NB_OF_TERMS constant

https://github.com/BruJu/wasmify-sophia/pull/10#discussion_r559373767\nThis constant was verbose and did not bring any meaningful information (the number 4 is already deeply tied with this class)

BruJu committed:
Ensemblist operations
